components/Buttons/src/MDCFloatingButton.m

```diff
@@ -60,79 +75,369 @@ - (instancetype)initWithFrame:(CGRect)frame shape:(MDCFloatingButtonShape)shape
+  if (shape == self.shape) {
+    [self updateShape];
+  }
+}
+
+- (void)setContentEdgeInsets:(UIEdgeInsets)contentEdgeInsets forShape:(MDCFloatingButtonShape)shape {
+  self.shapeToContentEdgeInsets[@(shape)] = [NSValue valueWithUIEdgeInsets:contentEdgeInsets];
+  if (shape == self.shape) {
+    [self updateShape];
+  }
+}
+
+- (void)setHitAreaInsets:(UIEdgeInsets)insets forShape:(MDCFloatingButtonShape)shape {
+  self.shapeToHitAreaInsets[@(shape)] = [NSValue valueWithUIEdgeInsets:insets];
+  if (shape == self.shape) {
+    [self updateShape];
+  }
+}
+
+- (void)setShape:(MDCFloatingButtonShape)shape {
+  _shape = shape;
+  [self updateShape];
+}
+
+- (BOOL)updateMinimumSize {
+  NSValue *sizeValue = self.shapeToMinimumSize[@(self.shape)];
+  if (!sizeValue && self.shape != MDCFloatingButtonShapeDefault) {
+    sizeValue = self.shapeToMinimumSize[@(MDCFloatingButtonShapeDefault)];
+  }
+
+  CGSize newSize = CGSizeZero;
+  if (sizeValue) {
+    newSize = [sizeValue CGSizeValue];
+  }
+
+  BOOL sizeChanged = (newSize.width > self.minimumSize.width)
+      || (newSize.height > self.minimumSize.height);
+  super.minimumSize = newSize;
+  if (sizeChanged) {
+    [self invalidateIntrinsicContentSize];
+  }
+  return sizeChanged;
+}
+
+- (BOOL)updateMaximumSize {
+  NSValue *sizeValue = self.shapeToMaximumSize[@(self.shape)];
+  if (!sizeValue && self.shape != MDCFloatingButtonShapeDefault) {
+    sizeValue = self.shapeToMaximumSize[@(MDCFloatingButtonShapeDefault)];
+  }
+
+  CGSize newSize = CGSizeZero;
+  if (sizeValue) {
+    newSize = [sizeValue CGSizeValue];
+  }
+
+  BOOL sizeChanged = !CGSizeEqualToSize(newSize, CGSizeZero)
+      && ((newSize.width < self.maximumSize.width) || (newSize.height < self.maximumSize.height));
+  super.maximumSize = newSize;
+  if (sizeChanged) {
+    [self invalidateIntrinsicContentSize];
+  }
+  return sizeChanged;
+}
+
+- (UIEdgeInsets)insetsForShape:(MDCFloatingButtonShape)shape {
+  NSValue *insetsValue = self.shapeToContentEdgeInsets[@(shape)];
+  if (!insetsValue && shape != MDCFloatingButtonShapeDefault) {
+    insetsValue = self.shapeToContentEdgeInsets[@(MDCFloatingButtonShapeDefault)];
+  }
+
+  if (insetsValue) {
+    return insetsValue.UIEdgeInsetsValue;
+  } else {
+    return UIEdgeInsetsZero;
+  }
+}
+
+- (void)updateContentEdgeInsets {
+  super.contentEdgeInsets = [self insetsForShape:self.shape];
+}
+
+- (void)updateHitAreaInsets {
+  NSValue *insetsValue = self.shapeToHitAreaInsets[@(self.shape)];
+  if (!insetsValue && self.shape != MDCFloatingButtonShapeDefault) {
+    insetsValue = self.shapeToHitAreaInsets[@(MDCFloatingButtonShapeDefault)];
+  }
+
+  if (insetsValue) {
+    super.hitAreaInsets = [insetsValue UIEdgeInsetsValue];
+  }
+}
+
+- (void)updateShape {
+  BOOL needsLayout = [self updateMinimumSize];
+  needsLayout |= [self updateMaximumSize];
+  [self updateContentEdgeInsets];
+  [self updateHitAreaInsets];
+  if (needsLayout) {
+    [self sizeToFit];
```

Comment:
This is concerning. I found that without a call to `sizeToFit` after UIAppearance applied the min/max sizes, the new sizes were not applied on-screen.  The alternative would be for all clients to do:

```objectivec
// Note that this is NOT |viewWillAppear:|
- (void)viewDidAppear:(BOOL)animated {
  [super viewDidAppear:animated];
  [self.floatingButton sizeToFit];
}
```